utils/s2n_random.c

```diff
@@ -346,7 +372,7 @@ int s2n_openssl_compat_rand(unsigned char *buf, int num)
     struct s2n_blob out = { 0 };
     POSIX_GUARD(s2n_blob_init(&out, buf, num));
 
-    if (s2n_result_is_error(s2n_get_private_random_data(&out))) {
+    if (s2n_result_is_error(s2n_get_custom_random_data(&out, &s2n_per_thread_rand_state.private_drbg))) {
```

Comment:
This is probably worth a comment. Without the context that this is never set for fips libcryptos, it looks wrong.

But what's the benefit of either making that "not fips" assertion at runtime or continuing to use s2n_get_private_random_data? Is there a performance concern?

Reply:
I added a comment. Both of your alternative suggestions are definitely possible.

I was hesitant to add a not-fips runtime assertion for the custom RAND callbacks because I think it's not possible to hit the assertion in a test. All of the RAND callback stuff is ifdef'd out if the libcrypto isn't FIPS, so a test wouldn't ever be able to attempt to set the callback and be in FIPS mode. But maybe this isn't a big deal? Would the assertion make more sense?

It's also possible to just continue using `s2n_get_private_random_data`. Since `S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND` is false for all FIPS libcryptos, `s2n_get_private_random_data` will always choose the custom random implementation whenever it's supported. However, if for some reason `S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND` is set incorrectly and FIPS mode is enabled when a custom callback is set, s2n-tls will segfault when generating random data due to infinite recursion (RAND_bytes -> custom callback -> RAND_bytes). I added a test to make sure `S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND` is set correctly in FIPS mode, but it still seemed more correct to call the custom implementation directly here.

Reply:
I reverted this change and added a check in `s2n_rand_init` to make sure the custom callbacks aren't set in FIPS mode. I also added a not-FIPS runtime check to `s2n_get_custom_random_data`.

Reply:
After thinking about this more, it seems like it might make more sense to put this in the OpenSSL RAND callback itself. The `S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND` flag should already prevent the RAND callback from being set on a FIPS libcrypto, and checking for not-fips directly before deciding between the libcrypto or custom implementation seems more safe. eeeb4c37a13e22ce5e27a54785e96efd63683044